### PR TITLE
Extend Device Plugin api to support resource attributes

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
**What this PR does / why we need it**:
This PR extends device plugin api object `Device` to include `capabilities`. This extension will be used as a building block for future items.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related #56649
Related: https://github.com/kubernetes/community/pull/782

**Release note**:

```release-note
None
```
/sig node
/area hw-accelerators
/cc  @jiayingz @RenaudWasTaken @vishh @ScorpioCPH @sjenning @derekwaynecarr @jeremyeder @lichuqiang @tengqm

### CODE REVIEW

**Comment by RenaudWasTaken** [comment]
> This PR extends device plugin api object Device to include capabilities. This extension will be used as a building block for future items.

What future items do you have in mind?
In general we might as well include the Kubelet side changes in this PR.

**Comment by vikaschoudhary16** [comment]
Resource classes

**Comment by RenaudWasTaken** [comment]
Can you explain how this ties to resource class ?

**Comment by vikaschoudhary16** [comment]
`capabilities` is the device metadata based on which resource class objects would be created. https://github.com/kubernetes/community/pull/782 should give an idea. Soon will be updating that PR.

File: pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go [hunk 1]
```diff
@@ -135,6 +135,8 @@ type Device struct {
 	ID string `protobuf:"bytes,1,opt,name=ID,json=iD,proto3" json:"ID,omitempty"`
 	// Health of the device, can be healthy or unhealthy, see constants.go
 	Health string `protobuf:"bytes,2,opt,name=health,proto3" json:"health,omitempty"`
+	// Capabilities of the device. eg: {"version": "1.0"}
+	Capabilities map[string]string `protobuf:"bytes,3,rep,name=capabilities" json:"capabilities,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
```
Comment (ScorpioCPH) [inline pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go]: Are these changes in this file auto-generated?
If yes, LGTM.

And maybe we can add some reference in PR description about how we auto-generate these code :)

**Review by ScorpioCPH — COMMENTED** [review]
@vikaschoudhary16 Thanks, I think this PR is helpful for `Resource Class`.

**Comment by RenaudWasTaken** [comment]
> capabilities is the device metadata based on which resource class objects would be created. kubernetes/community#782 should give an idea. Soon will be updating that PR.

I'm wondering if `capabilities` is the right name as it might be confused with linux capabilities.

**Comment by jeremyeder** [comment]
agree renaud, something like hwattr maybe? or hwcapabilities?

**Comment by vikaschoudhary16** [comment]
Since `capabilities` is a member of `Device` structure, so it looks to me that it should be clear enough that these are capabilities of this same device. Anyways, how about `DeviceProperties` ?

**Comment by ScorpioCPH** [comment]
+1 for `Properties`.

File: pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go (hunk 1 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go]: yes, this is auto-generated. And here is how to generate: 
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/deviceplugin/v1alpha/api.proto#L1

**Comment by RenaudWasTaken** [comment]
> Since capabilities is a member of Device structure, so it looks to me that it should be clear enough that these are capabilities of this same device.

Take the annotation example in the case of containers. When you see annotations in a sentence it's not always easy to spot whether someone is talking about K8s annotations, CRI annotations, Docker annotations or OCI annotations.
Having a name that doesn't need to much context might avoid confusions (which already happened to me at the beginning of this issue).

> Anyways, how about DeviceProperties ? or DeviceAttributes ?

I think K8s / golang conventions are that you shouldn't repeat the structure name ?
`Properties` or `Attributes` sounds good though :)

File: pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go [hunk 2]
```diff
@@ -135,6 +135,8 @@ type Device struct {
 	ID string `protobuf:"bytes,1,opt,name=ID,json=iD,proto3" json:"ID,omitempty"`
 	// Health of the device, can be healthy or unhealthy, see constants.go
 	Health string `protobuf:"bytes,2,opt,name=health,proto3" json:"health,omitempty"`
+	// Attributes of the device. eg: {"version": "1.0"}
+	Attributes map[string]string `protobuf:"bytes,3,rep,name=attributes" json:"attributes,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
```
Comment (resouer) [inline pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go]: what are these information come from. One use case in my mind is querying nvidia-plugins API. Am I right?

Comment (vikaschoudhary16) [reply pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go]: Numa manager can also use this:
https://github.com/kubernetes/community/pull/1680#discussion_r163776284

Comment (vishh) [reply pkg/kubelet/apis/deviceplugin/v1alpha/api.pb.go]: The overall design for device topology awareness is not yet finalized. Let's hold off on this PR for now.

**Comment by resouer** [comment]
@vishh @RenaudWasTaken Just a follow up here.

 A solid use case for this proposal would be topology of hardware accelerators. Having only a `DeviceID` to describe a device is now the main blocker for us to use device plugin.

**Pull request closed without merging**